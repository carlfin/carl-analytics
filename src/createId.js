import cookies from 'js-cookie';
import { v4 as uuidv4, v5 as uuidv5 } from 'uuid';

const KEY = 'carlId';
const NAMESPACE = '1b671a64-40d5-491e-99b0-da01ff1f3341';

function createId() {
  // Try to read a `carlId` cookie value directly.
  const cookieVal = cookies.get(KEY);

  // Return existing value.
  if (cookieVal) {
    return cookieVal;
  }

  // Read the id from the Google Analytics cookie.
  const gaCookie = cookies.get('_ga');

  // Generate a unique id. If there is a _ga value read then use this as a base.
  const id = gaCookie ? uuidv5(gaCookie, NAMESPACE) : uuidv4();

  // Set a new cookie with the generated value.
  cookies.set(KEY, id);

  return id;
}

export default createId;
