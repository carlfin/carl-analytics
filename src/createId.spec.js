import cookies from 'js-cookie';
import createId from './createId';

jest.mock('js-cookie', () => ({
  get: jest.fn(),
  set: jest.fn(),
}));

describe('createId()', () => {
  beforeEach(() => {
    cookies.get.mockReset();
  });

  it('should return an existing value', () => {
    cookies.get.mockImplementation(() => 'abcdef');
    expect(createId()).toBe('abcdef');
  });

  it('should return value based on `_ga`', () => {
    cookies.get.mockImplementation((key) => {
      const values = {
        _ga: 'GA-1234',
      };

      return values[key];
    });

    const id = createId();

    expect(id).toMatchSnapshot();
    expect(cookies.set).toHaveBeenCalledWith('carlId', id, { expires: 365 });
  });

  it('should return a new random value', () => {
    const id = createId();

    expect(id).toEqual(expect.any(String));
    expect(cookies.set).toHaveBeenCalledWith('carlId', id, { expires: 365 });
  });
});

