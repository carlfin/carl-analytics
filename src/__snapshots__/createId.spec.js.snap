// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createId() should return value based on \`_ga\` 1`] = `"487260d5-84da-5b5a-86c1-c6539321c08e"`;
